src/mlpack/methods/ann/vanilla_rbm.hpp

```diff
@@ -0,0 +1,207 @@
+/**
+ * @file vanilla_rbm.hpp
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_VANILLA_RBM_HPP
+#define MLPACK_METHODS_ANN_VANILLA_RBM_HPP
+
+#include <mlpack/core.hpp>
+#include <mlpack/prereqs.hpp>
+#include <mlpack/core/math/random.hpp>
+
+#include "layer/layer.hpp"
+#include "layer/base_layer.hpp"
+
+#include "activation_functions/softplus_function.hpp"
+#include "init_rules/gaussian_init.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InitializationRuleType,
+    typename VisibleLayerType,
+    typename HiddenLayerType>
+class RBM
+{
+ public:
+  using NetworkType = RBM<InitializationRuleType, VisibleLayerType,
+      HiddenLayerType>;
+
+  /* 
+   * Intalise all the parameters of the network
+   * using the intialise rule. 
+   *
+   * @tparam: IntialiserType rule to intialise the parameters of the network
+   * @tparam: VisibleLayerType visible layer 
+   * @tparam: HiddenLyaerType hidden layer
+   */
+  RBM(arma::mat predictors, InitializationRuleType initializeRule,
+      VisibleLayerType visible,
+      HiddenLayerType hidden,
+      const size_t numSteps = 1,
+      const bool persistence = false);
+
+  // Reset the network
+  void Reset();
+
+  /* 
+   * Train the netwrok using the Opitimzer with given set of args.
+   * the optimiser sets the paratmeters of the network for providing
+   * most likely parameters given the inputs
+   * @param: predictors data points
+   * @param: optimizer Optimizer type
+   * @param: Args arguments for the optimizers
+   */
+  template<template <typename> class Optimizer>
+  void Train(const arma::mat& predictors, Optimizer<NetworkType>& optimizer);
+
+ /**
+  * Evaluate the rbm network with the given parameters. This function
+  * is usually called by the optimizer to train the model.
+  * The function computes the pseudo likelihood
+  *
+  * @param parameters Matrix model parameters.
+  * @param i Index of point to use for objective function evaluation.
+  * 
+  */
+  double Evaluate(const arma::mat& parameters, const size_t i);
+
+ /** 
+  * This function calculates
+  * the free energy of the model
+  * @param: input data point 
+  */
+  double FreeEnergy(arma::mat&& input);
```

Comment:
Why do you use rvalue references here? As for me the input should be`const arma::mat&` or even a template since you calculate free energy of `arma::vec`.